[AMDGPU] Replace `isInlinableLiteral16` with specific version

shiltian opened the pull request:
The current implementation of `isInlinableLiteral16` assumes, a 16-bit inlinable
literal is either an `i16` or a `fp16`. This is not always true because of
`bf16`. However, we can't tell `fp16` and `bf16` apart by just looking at the
value. This patch splits `isInlinableLiteral16` into three versions, `i16`,
`fp16`, `bf16` respectively, and call the corresponding version.

rampitec commented on llvm/test/CodeGen/AMDGPU/wmma-gfx12-w32-imm.ll:
```diff
@@ -112,7 +112,7 @@ bb:
 ; GFX12-LABEL: test_wmma_bf16_16x16x16_bf16_imm:
 ; GFX12:       ; %bb.0: ; %bb
-; GFX12-NEXT:    v_wmma_bf16_16x16x16_bf16 v[10:13], v[0:3], v[4:7], 1.0
+; GFX12-NEXT:    v_wmma_bf16_16x16x16_bf16 v[10:13], v[0:3], v[4:7], 0
```
What has happened here again? I have not seen this downstream. I would understand a code bloat with the initialization with 0x3f80, but not outright 0 inline.

Also just as a note pulled from downstream: this shall use BF16 in the profile. As a minimum in the profile, but then also in the intrinsic. But downstream it still maintains 1.0, so something is missing.

shiltian replied:
> Also just as a note pulled from downstream: this shall use BF16 in the profile. As a minimum in the profile, but then also in the intrinsic. But downstream it still maintains 1.0, so something is missing.

Yes. That is on the top of my list. I'll fix the downstream today.

piotrAMD replied:
As for the intrinsic v_wmma_bf16_16x16x16_bf16, changing it is tricky because the same one is being used on older gens. The proposal (https://github.com/llvm/llvm-project/pull/77795#pullrequestreview-1839026480) was to add a new one with proper bf16 type for use on gfx12.

shiltian replied:
Thanks for the information. I'll see what we can do here.

arsenm replied:
No, the proposal was to change all the intrinsics. We only need to worry about maintaining compatibility with the C builtins for old operations

shiltian replied:
If that is the case, I think https://github.com/llvm/llvm-project/pull/81669 can meet the requirement.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:
```diff
@@ -3327,35 +3327,41 @@ bool AMDGPUDAGToDAGISel::SelectWMMAVISrc(SDValue In, SDValue &Src) const {
   // 16 bit splat
   SDValue SplatSrc32 = stripBitcast(In);
-  if (auto *SplatSrc32BV = dyn_cast<BuildVectorSDNode>(SplatSrc32)) {
+  if (auto *SplatSrc32BV = dyn_cast<BuildVectorSDNode>(SplatSrc32))
```
For some reason this trunk of code was missing in the previous version. That caused the two WMMA instructions to be generated incorrectly. Good catch!

rampitec replied:
Kudos!

shiltian committed:
[ADMGPU] Replace `isInlinableLiteral16` with specific version

The current implementation of `isInlinableLiteral16` assumes, a 16-bit inlinable
literal is either an `i16` or a `fp16`. This is not always true because of
`bf16`. However, we can't tell `fp16` and `bf16` apart by just looking at the
value. This patch splits `isInlinableLiteral16` into three versions, `i16`,
`fp16`, `bf16` respectively, and call the corresponding version.

rampitec reviewed (approved):
LGTM
